grype/presenter/sarif/presenter.go

```diff
@@ -0,0 +1,348 @@
+	var path string
+	if len(m.Package.Locations) > 0 {
+		path = m.Package.Locations[0].RealPath
+	}
+
+	switch pres.srcMetadata.Scheme {
+	case source.DirectoryScheme:
+		if pres.srcMetadata.Path != "" {
+			return strings.TrimPrefix(fmt.Sprintf("%s/%s", pres.srcMetadata.Path, path), "./")
+		}
+	case source.FileScheme:
+		if pres.srcMetadata.Path != "" {
+			return pres.srcMetadata.Path
+		}
+	}
+
+	return path
+}
+
+// locations the locations array is a single "physical" location with potentially multiple logical locations
+func (pres *Presenter) locations(m match.Match) []*sarif.Location {
+	var logicalLocations []*sarif.LogicalLocation
+	physicalLocation := pres.location(m)
+
+	switch pres.srcMetadata.Scheme {
+	case source.ImageScheme:
+		img := pres.srcMetadata.ImageMetadata.UserInput
+		for _, l := range m.Package.Locations {
+			logicalLocations = append(logicalLocations, &sarif.LogicalLocation{
+				FullyQualifiedName: sp(fmt.Sprintf("%s@%s:%s", img, l.FileSystemID, l.RealPath)),
+				Name:               sp(l.RealPath),
+			})
+		}
+
+		// this is a hack to get results to show up in GitHub, as it requires relative paths for the location
+		// but we really won't have any information about what Dockerfile on the filesystem was used to build the image
+		// TODO we could add configuration to specify the prefix, a user might want to specify an image name and architecture
+		// in the case of multiple vuln scans, for example
+		physicalLocation = fmt.Sprintf("image/%s", strings.TrimPrefix(physicalLocation, "/"))
+	case source.FileScheme:
+		for _, l := range m.Package.Locations {
+			logicalLocations = append(logicalLocations, &sarif.LogicalLocation{
+				FullyQualifiedName: sp(fmt.Sprintf("%s:%s", pres.srcMetadata.Path, l.RealPath)),
+				Name:               sp(l.RealPath),
+			})
+		}
+
+		physicalLocation = pres.srcMetadata.Path
+	}
+
+	return []*sarif.Location{
+		{
+			PhysicalLocation: &sarif.PhysicalLocation{
+				ArtifactLocation: &sarif.ArtifactLocation{
+					URI: sp(physicalLocation),
+				},
+				// TODO When grype starts reporting line numbers this will need to get updated
+				Region: &sarif.Region{
+					StartLine:   ip(1),
+					StartColumn: ip(1),
+					EndLine:     ip(1),
+					EndColumn:   ip(1),
+				},
+			},
+			LogicalLocations: logicalLocations,
+		},
+	}
+}
+
+// severityText provides a textual representation of the severity level of the match
+func (pres *Presenter) severityText(m match.Match) string {
+	meta := pres.metadata(m)
+	if meta != nil {
+		severity := vulnerability.ParseSeverity(meta.Severity)
+		switch severity {
+		case vulnerability.CriticalSeverity:
+			return "critical"
+		case vulnerability.HighSeverity:
+			return "high"
+		case vulnerability.MediumSeverity:
+			return "medium"
+		}
+	}
+	return "low"
+}
+
+// securitySeverityValue GitHub uses a numeric severity value to determine whether things are critical, high, etc.
+func (pres *Presenter) securitySeverityValue(m match.Match) string {
```

Comment:
Are these numbers expected to correspond to the CVSS scale? The number range looks similar so I wanted to check, in case we should handle this especially for when CVSS data is present

Reply:
This corresponds to the GitHub security-severity property (as noted in the comment 😄)
> Recommended. A string representing a score that indicates the level of severity, between 0.0 and 10.0, for security queries (tags includes security). This, with the properties.precision property, determines whether the results are displayed by default on GitHub so that the results with the highest security-severity, and highest precision are shown first. Code scanning translates numerical scores as follows: over 9.0 is critical, 7.0 to 8.9 is high, 4.0 to 6.9 is medium and 3.9 or less is low.

Although there is a typo, it should say "9.0 and greater" or similar

Reply:
Added a little more info to the comment, mentioning the `security-severity` property

Reply:
So actually these numbers correspond **exactly** with CVSS v3.0 ratings. (See this if you're unfamiliar with these ratings: https://nvd.nist.gov/vuln-metrics/cvss.)

I'm wondering if it makes sense to use our CVSS data when we have it, to preserve some precision about the assigned severity? Just a thought

Reply:
Ah yeah, makes sense; I'll also add this context about why we are returning CVSS if we have it :)